certbot-apache/certbot_apache/_internal/prefetch_ocsp.py

```diff
@@ -126,6 +127,29 @@ def _ocsp_refresh(self, cert_path, chain_path):
             logger.warning("Encountered an issue while trying to prefetch OCSP "
                            "response for certificate: %s", cert_path)
 
+    def _ocsp_ttl(self, next_update):
+        """Calculates Apache internal TTL for the next OCSP staple
+        update.
+
+        The resulting TTL is half of the time between now
+        and the time noted by nextUpdate field in OCSP response.
+
+        If nextUpdate value is None, a default value will be
+        used instead.
+
+        :param next_update: datetime value for nextUpdate or None
+
+        :returns: TTL in seconds.
+        :rtype: int
+        """
+
+        if next_update:
+            now = time.time()
+            res_ttl = int(time.mktime(next_update.timetuple()) - now)
+            if res_ttl > 0:
+                return res_ttl/2
```

Comment:
`nextUpdate` is the latest time when this OCSP staple will be valid, and is basically promising "an updated OCSP response will be available before this time". 

Apache uses 1 hour TTL internally, so it'll attempt to refresh the staple every hour, triggered by incoming request. We want to prevent this hitting basically ever as Apache does poor job in validating the responses or retrying before replacing the existing response.

For Let's Encrypt the `nextUpdate` is 7 days from `thisUpdate`, and I'm not 100% sure about when the new one will be available, but it has been days in my tests.

I chose half of the time between now and `nextUpdate` as TTL value as it will scale, while it should never be low enough for Apache's internal response fetching to trigger.

Reply:
What do you mean by it will scale?

To me personally it seems like the value we use should be as close to the time in `nextUpdate` as we can make it while being confident that Apache will refresh the response before it expires. This delays Apache refreshing the value for as long as possible and gives Certbot more time to do so. Do you see any problems with this approach?

If we took that approach though, I'm still not sure what this value should be. The TTL value we're putting in the DBM file is a duration rather than a time of when it expires. Do you know when Apache reads this value and starts the clock ticking towards its expiration?

Reply:
> What do you mean by it will scale?

I was thinking that regardless of the time delta between now and `nextUpdate`, we'd always be writing a safe value in the DBM.

> To me personally it seems like the value we use should be as close to the time in nextUpdate as we can make it while being confident that Apache will refresh the response before it expires.

Yeah, you are most likely right here. To me it feels like the proper value would be something in the ballpark of `nextUpdate` - 12h. I'll make this change.

> The TTL value we're putting in the DBM file is a duration rather than a time of when it expires.

The final value we write to DBM is actually a fully fledged timestamp. It's handled in [apache_util.get_apache_ocsp_struct()](https://github.com/certbot/certbot/blob/ocsp_apache_continued/certbot-apache/certbot_apache/_internal/apache_util.py#L11-L24). Based on this discussion I'm thinking we probably should make it more apparent from the caller as well.